### PR TITLE
Performance tracker

### LINKED ISSUE #8847: Performance tracker memory leak
There might be a memory leak in the attestation performance tracker

This happens systematically at every epoch transition, righ after sync committee pref log
image

### PR DESCRIPTION
**Pull request by mehdi-aouadi**
## PR Description
Update the performance tracker to support the new Electra attestation aggregation logic.
This prevents a memory leak triggered in Electra.

## Fixed Issue(s)

Fixes #8847 

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java [hunk 1]
```diff
@@ -300,16 +301,23 @@ private AttestationPerformance calculateAttestationPerformance(
...
-    final Map<Bytes32, NavigableMap<UInt64, SszBitlist>> slotAndBitlistsByAttestationDataHash =
-        new HashMap<>();
+    final Map<Bytes32, NavigableMap<UInt64, AttestationBitsAggregator>>
+        slotAndBitlistsByAttestationDataHash = new HashMap<>();
     for (Map.Entry<UInt64, List<Attestation>> entry : attestationsIncludedOnChain.entrySet()) {
+      final Optional<Int2IntMap> committeesSize =
+          Optional.of(spec.getBeaconCommitteesSize(state, entry.getKey()));
       for (Attestation attestation : entry.getValue()) {
         final Bytes32 attestationDataHash = attestation.getData().hashTreeRoot();
-        final NavigableMap<UInt64, SszBitlist> slotToBitlists =
```
Comment (rolfyone) [inline beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java:308]: we could potentially use JMH on this...

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java [hunk 2]
```diff
@@ -314,9 +317,9 @@ private AttestationPerformance calculateAttestationPerformance(
         slotToBitlists.merge(
             entry.getKey(),
             AttestationBitsAggregator.of(attestation, committeesSize),
-            (first, second) -> {
-              first.or(second);
-              return first;
+            (firstBitsAggregator, secondBitAggregator) -> {
+              firstBitsAggregator.or(secondBitAggregator);
```
Comment (tbenr) [inline beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java]: I think here it make sense to let `or` return `this`... so we can chain operations and also have a compact code here..

Comment (tbenr) [reply beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java]: but we can leave it as a separate PR

Comment (tbenr) [reply beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java]: thinking about this a second time, I let the method return void on purpose to be clear that the object mutates.
I think it is safer this way, so ignore me.

Comment (mehdi-aouadi) [reply beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java]: The change is tiny and makes it more readable/concise, I added it :)

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java (hunk 1 above)
Comment (mehdi-aouadi) [reply beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java:308]: Doing that, will share the results

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java (hunk 2 above)
Comment (mehdi-aouadi) [reply beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java]: Just saw your last comment. Well, ok. Reverted it

**Comment by mehdi-aouadi** [comment]
I run some benchmarks to compare the old and the new attestations performance algorithm and the results are encouraging.

**Old Version**

Phase0 Attestations
```
Benchmark                                                    (attestationsCount)   Mode  Cnt      Score     Error  Units
PerformanceTrackerBenchmark.calculateAttestationPerformance                  100  thrpt   10  27119.554 ± 597.657  ops/s
PerformanceTrackerBenchmark.calculateAttestationPerformance                 1000  thrpt   10   2558.557 ±  99.012  ops/s
PerformanceTrackerBenchmark.calculateAttestationPerformance                10000  thrpt   10    161.646 ±   1.570  ops/s
PerformanceTrackerBenchmark.calculateAttestationPerformance               100000  thrpt   10     10.397 ±   0.026  ops/s
```
Doesn't work for the electra attestations, so no Electra attestations benchmarking
____________

**New Version**

Phase0 Attestations
```
Benchmark                                                    (attestationsCount)   Mode  Cnt      Score     Error  Units
PerformanceTrackerBenchmark.calculateAttestationPerformance                  100  thrpt   10  26828.725 ± 410.393  ops/s
PerformanceTrackerBenchmark.calculateAttestationPerformance                 1000  thrpt   10   2638.513 ±  34.554  ops/s
PerformanceTrackerBenchmark.calculateAttestationPerformance                10000  thrpt   10    160.236 ±   2.944  ops/s
PerformanceTrackerBenchmark.calculateAttestationPerformance               100000  thrpt   10     11.682 ±   0.054  ops/s
```

Electra Attestations
```
Benchmark                                                    (attestationsCount)   Mode  Cnt      Score     Error  Units
PerformanceTrackerBenchmark.calculateAttestationPerformance                  100  thrpt   10  13805.354 ± 120.632  ops/s
PerformanceTrackerBenchmark.calculateAttestationPerformance                 1000  thrpt   10   1733.083 ±   8.922  ops/s
PerformanceTrackerBenchmark.calculateAttestationPerformance                10000  thrpt   10     98.000 ±   0.780  ops/s
PerformanceTrackerBenchmark.calculateAttestationPerformance               100000  thrpt   10      9.142 ±   0.030  ops/s
```
____________

Context:
- The operations refer to the `calculateAttestationPerformance` algorithm
- The `attestationsCount` is the number of attestations being processed. They are all attesting from the same slot and hence all aggregated (which is extreme)

Observations:
- The new algorithm performance is almost the same for the phase0 attestations
- The Electra attestations require almost twice the processing time but that's normal since the aggregation bit data structure is more complex. If we take into consideration that there are less attestations in Electra (more aggregation) this should be fine

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java (hunk 1 above)
Comment (mehdi-aouadi) [reply beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java:308]: Done: https://github.com/Consensys/teku/pull/8851#issuecomment-2506724226

**Review by rolfyone — COMMENTED** [review]
LGTM

**Review by rolfyone — APPROVED** [review]

**Comment by rolfyone** [comment]
we'll need to cherrypick this to the mekong branch too just remember...

**Pull request merged**